### DISCUSSION TITLE
[Application Development] Farcaster Frame Image Response Timed Out

### DISCUSSION THREAD

**Discussion by mubarakone** [ethereum-optimism/developers#345] [General]
### Did you check the documentation?

- [X] Yes

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

This is just a duplicate from an issue opened in onchainkit repo from Coinbase. There doesn't seem to be too many people working on that repo, so I am going to post that issue here for more eyeballs:
https://github.com/coinbase/onchainkit/issues/266

**Comment by sbvegan** [comment]
Hey @mubarakone, I'm not too familiar with building frames, but I know some builders in the space. I'll see if they can hop in and help you out here!

**Comment by jakim929** [comment]
Cool frame! thx for providing the repo + lots of context

Seems like there's two problems
- Farcaster frame handlers need to return within 5s or they time out
- Nextjs node runtime has 10s limit on request duration

I did a quick performance benchmark of your your fetch function like this and it seems like the majority of time is taken for the following tasks. The code I used to benchmark is [here](https://gist.github.com/jakim929/29cde5b11bd4875161a870123a14d85a)

1. generating chatGPT summary - 5 ~ 8s
2. generating & saving image ~ 2s

```

[!] Time taken to readFileAndGenerateSummary:  7988.090124964714
		Time taken to fetch MDX content:  96.4855409860611
		Time taken to generate summary using GPT:  7890.897707939148
[!] Time taken for all generateAndSaveImage:  2566.829041957855 - parallelized
		Time taken to fetch image from og for index 2:  1066.7457920312881
		Time taken to fetch image from og for index 5:  1099.7112910747528
		Time taken to fetch image from og for index 3:  1117.6078330278397
		Time taken to fetch image from og for index 1:  1176.9544169902802
		Time taken to fetch image from og for index 0:  1746.1630419492722
		Time taken to uploadBytes for index 1:  698.3835000991821
		Time taken to uploadBytes for index 5:  775.2114169597626
		Time taken to fetch image from og for index 4:  1885.2350420951843
		Time taken to uploadBytes for index 2:  631.8874160051346
		Time taken to uploadBytes for index 0:  651.742124915123
		Time taken to uploadBytes for index 4:  634.2836250066757
		Time taken to uploadBytes for index 3:  668.6177920103073
[!!] Time taken for the entire response:  10556.415666937828
```


Given that the timeout for Frame requests (before Warpcast client considers it failed) is ~ 5000ms, in general you want your response to be under 2s max, since there's usually random variance in response time as well.



**It's going be very difficult to bring this down to a reasonable response time (especially the chatGPT call), so I suggest you**
**1) set up a loading screen when a user is attempting to generate a summary for the first time,** 
**2) then pass off the generation task to a different handler.**

### Loading + refresh flow

when someone requests a summary
- check db / bucket to see if summary/images have already been generated
	- if yes, return the images immediately
	- if no
		- trigger the chatgpt + image generation + save file flow on a separate handler
		- then return a loading screen + refresh button

when someone presses refresh
- check db / bucket to see if summary/images have already been generated
	- if yes, return the images immediately
	- if no
		- then return a loading screen + refresh button

### Separate handler

You will probably need some other endpoint / worker that will be in charge of calling chatgpt then uploading the images. Some thoughts
- if you're using nextjs+vercel, use the edge runtime https://vercel.com/docs/functions/runtimes#size-limits since it will allow for up to 25s. the chatgpt call takes ~7s, so you probably don't want to push it.
- you probably want a job queue type pattern like this https://upstash.com/docs/qstash/quickstarts/vercel-nextjs - I haven't used upstash but just one option

**Comment by mubarakone** [comment]
Thanks for the swift response

If I just moved everything inside `getResponse()` to a separate function in the same file, and only have `getResponse()` wait for a boolean variable to be true, will the server stop generating the images and refresh the server? Starting everything over?

```typescript
let finalImageSaved = false;   // Have the boolean set to false by default

async function generateAndSaveImage(textsnippet, index, storage) {

      // Once all images are uploaded, set finalImageSaved = true

      if ((index == 6) && snapshot) {
          finalImageSaved = true;
      }
}

// New function

async function generateFrame(): Promise<void> {

     // Move everything inside getResponse() to here
    
}

async function getResponse(req: NextRequest): Promise<NextResponse> {
 
  //  Only if finalImageSaved is false

  if (!finalImageSaved) {
  
      return new NextResponse(
        getFrameHtmlResponse({ 
          buttons: [
            {
              action: 'post',
              label: 'Refresh 🔄',
            },
          ],
          image: {
            src: `https://newspaper.tips/loading-image.png`,
            aspectRatio: '1:1',
          },
          postUrl: `https://newspaper.tips/api/generateFrames`,  //  <--- This will call this endpoint in a recursive loop, until finalImageSaved returns true
         }),

  }

  return new NextResponse(
    getFrameHtmlResponse({ 
      buttons: [
        {
          action: 'post',
          label: '⬅️ Back',
        },
        {
          label: `1/6`,
        },
        {
          action: 'post',
          label: 'Next ➡️',
        },
      ],
      image: {
        src: 'http://34.36.130.28/images/snippet_0.png',
        aspectRatio: '1:1',
      },
      postUrl: `https://newspaper.tips/api/frame`,
    }),
  );
}

export async function POST(req: NextRequest): Promise<Response> {
  return getResponse(req);
}

```

I'm thinking of having the loading screen + refresh button return in the same function `getResponse()`, and just wait for a boolean after all the images have been uploaded. Maybe that is what you meant by having a job queue.

**Reply by jakim929** [reply]
I don't think that will work 
1. **the generateFrame call won't finish**: If you don't `await` the `generateFrame()` in your `getResponse()` handler, the nextjs serverless function will return the response then kill the unawaited `generateFrame()` call. But even if you were able to await it, it will exceed the execution duration limit 10s for a nextjs handler
2. **it will run the same job multiple times**: a user can repeatedly refresh lots of times, and each one will cause a call to chatGPT. not great since the API is pretty expensive (~ cents per call)

by job queue I meant something like this 
https://upstash.com/docs/qstash/features/background-jobs
not sure if nextjs/vercel has something native that does this though

what you want is some way to 
1. run the generateFrame() **once**
2. make sure the job handler finishes running the full flow (generating summary, creating images, then uploading)

**Reply by mubarakone** [reply]
Sorry but middleware is my weakness, so I am going to need extra help.
This is my progress so far: https://github.com/mubarakone/tailwind-css-starter-blog/tree/main/app/api

The `startBackgroundJob` gets called, but QStash URL endpoint doesn't pass through `/api/handleGenerateFrames` for some reason. The Generating Summary image refresh button is able to call `handleCallback` endpoint.
Test the frame yourself if you have time, using the same link I provided earlier
